pubsub/redis.go

```diff
@@ -38,13 +42,79 @@ func NewRedisSubscriber(node *node.Node, url string, channel string) RedisSubscr
 // Start connects to Redis and subscribes to the pubsub channel
 func (s *RedisSubscriber) Start() error {
 	// Check that URL is correct first
-	_, err := url.Parse(s.url)
+	redisUrl, err := url.Parse(s.url)
 
 	if err != nil {
 		return err
 	}
 
+	var sntnl *sentinel.Sentinel
+	var password string
+
+	if s.sentinels != "" {
+		masterName := redisUrl.Hostname()
+		password, _ = redisUrl.User.Password()
+
+		s.log.Debug("Redis sentinel enabled")
+		s.log.Debugf("Redis sentinel parameters:  sentinels: %s,  masterName: %s", s.sentinels, masterName)
+		sentinels := strings.Split(s.sentinels, ",")
+		sntnl = &sentinel.Sentinel{
+			Addrs:      sentinels,
+			MasterName: masterName,
+			Dial: func(addr string) (redis.Conn, error) {
+				timeout := 500 * time.Millisecond
+
+				c, err := redis.Dial(
+					"tcp",
+					addr,
+					redis.DialConnectTimeout(timeout),
+					redis.DialReadTimeout(timeout),
+					redis.DialReadTimeout(timeout),
+				)
+				if err != nil {
+					s.log.Debugf("Failed to connect to sentinel %s", addr)
+					return nil, err
+				}
+				s.log.Debugf("Successfully connected to sentinel %s", addr)
+				return c, nil
+			},
+		}
+
+		defer sntnl.Close()
+
+		// Periodically discover new Sentinels.
+		go func() {
+			err := sntnl.Discover()
+			if err != nil {
+				s.log.Warn("Failed to discover sentinels")
+			}
+			for {
+				<-time.After(30 * time.Second)
+				err := sntnl.Discover()
+				if err != nil {
+					s.log.Warn("Failed to discover sentinels")
+				}
+			}
+		}()
+	}
+
 	for {
+
+		if s.sentinels != "" {
+			masterAddress, err := sntnl.MasterAddr()
+
+			if err != nil {
+				s.log.Warn("Failed to get master address from sentinel.")
+				return err
+			}
+			s.log.Debugf("Got master address from sentinel: %s", masterAddress)
+
+			if password == "" {
```

Comment:
I don't get the idea of reconstruction `s.url` if the propper URL is already there. Especially, doing that we loose url schema which can contain `rediss` for TLS-secured instances.

Reply:
What do you mean with: the proper url is already there? We go from `redis://[:pass@]<mastername>` to `redis://[:pass@]<actual ip of the current master>`.  Didn't know scheme can change. I'll use the pased one here instead of hadcoding it.

Reply:
or should we just replace host of the redisUrl object and use string here? guess thats what you meant

Reply:
did that